File: flask/src/validation/validation.py
from pydantic import ValidationError
from response.response_error import ResponseError

# Fungsi untuk melakukan validasi data menggunakan schema_class
def validate(schema_class, data):
    try:
        # Mencoba untuk memvalidasi data sesuai dengan schema_class
        return schema_class(**data)
    except ValidationError as e:
        # Jika terjadi error validasi, mengambil pesan error dari setiap masalah yang ditemukan
        for error in e.errors():
            message = error['msg']
        
        # Menaikkan ResponseError jika terjadi error validasi
        raise ResponseError(400, message=message)

File: express/src/validation/validation.js
import { ResponseError } from "../response/response-error.js";

// Fungsi untuk melakukan validasi menggunakan schema yang telah didefinisikan
const validate = (schema, request) => {
  // Melakukan validasi pada request menggunakan schema
  const result = schema.validate(request, {
    abortEarly: false, // Melanjutkan validasi meskipun ada kesalahan
    allowUnknown: false, // Menolak field yang tidak dikenali dalam request
  });

  // Jika ada error, lemparkan ResponseError dengan status 400 dan pesan error
  if (result.error) {
    throw new ResponseError(400, result.error.message);
  } else {
    // Kembalikan data yang sudah divalidasi
    return result.value;
  }
};

export { validate };
